[bugfix] Modernize

trombik opened the pull request:
this PR fixes bad practices in the role and its tests while trying NOT to change the behaviors. there are still issues, but to fix them require overhauling which I do avoid in this PR.

Commit:
QA

Commit:
normalize `when`

Commit:
rewrite unit test

and use standard `.kitchen.yml`

Commit:
patch rc.d/namenode with lineinfile

and remove reallyenglish.freebsd-repos

Commit:
introduce hadoop_namenode_host and hadoop_namenode_port

Commit:
[bugfix] use `insertafter` to patch the file

also, enable the service here (necessary to support OpenBSD)

Commit:
ensure the service is stopped

and replace deprecated command `hadoop namenode -format`.

remove `enabled: true` from `tasks/main.yml`.

Commit:
fix up specs

_raw_ path without `file://` is deprecated.

also, catch up with a change, fs.defaultFS, in the playbook

Commit:
remove pre_tasks

because the startup script is now patched by the role.

also:

* rename s/ansible-role-hosts/reallyenglish.hosts/g
* use canonical path to `hadoop.tmp.dir`

Commit:
update Rakefile and add requirements.yml

Commit:
enable pipelining

Commit:
QA

Commit:
enable pipelining

Commit:
QA

Commit:
use dest for backward-compatibility

Commit:
use requirements.yml

Commit:
set hadoop_namenode_host to the binded interface

remove freebsd_repos_url

Commit:
fix invalid name

Commit:
remove extra attributes from requirements.yml

zookeeper has been released

Commit:
Merge branch 'master' into modernize

Commit:
QA

Commit:
QA

remove unused code

Commit:
update README

Commit:
retry at least three times before giving up

Commit:
typo

Commit:
use succeeded filter instead of gazing the darkness in registered

variable

Commit:
increase FETCH_TIMEOUT

although the man page does not state what timeout it is, the only option
that can be changed to solve the issue is this one.

Commit:
apply the workaround to the integration tests

FETCH_TIMEOUT does not always workaround the timeout issue, though.

trombik commented:
the PR is now ready for review as three consecutive builds have been succeeded

mitsururike commented on tasks/configure-xml.yml:
```diff
@@ -44,4 +52,4 @@
 - set_fact:
     hadoop_config_changed: true
-  when: "{{ core_site.changed or hdfs_site.changed or mapred_site.changed or yarn_site.changed or (hadoop_config_changed is defined and hadoop_config_changed) }}"
+  when: core_site.changed or hdfs_site.changed or mapred_site.changed or yarn_site.changed or (hadoop_config_changed is defined and hadoop_config_changed)
```
loop should be used for repetition. the code below worked.
```
---

- name: "Configure {{ item.file }}"
  template:
    src: "{{ item.name }}-site.xml.j2"
    dest: "{{ item.file }}"
    validate: "xmllint %s"
  when:
    - item.file is defined
    - item.file | length > 0
  register: hadoop_site
  notify:
    - Restart namenode
  with_items:
    - { name: 'core', file: "{{ hadoop_core_site_file }}" }
    - { name: 'hdfs', file: "{{ hadoop_hdfs_site_file }}" }
    - { name: 'mapred', file: "{{ hadoop_mapred_site_file }}" }
    - { name: 'yarn', file: "{{ hadoop_yarn_site_file }}" }

# XXX restart hadoop services when hadoop_config_changed is true
# you cannot "notify" here because zkfc is not enabled when standalone
- set_fact:
    hadoop_config_changed: true
  when: hadoop_site.changed or (hadoop_config_changed is defined and hadoop_config_changed)
```

mitsururike commented on tasks/install-FreeBSD.yml:
```diff
@@ -15,3 +21,16 @@
+  # XXX service namenode status has been broken
+  lineinfile:
+    dest: /usr/local/etc/rc.d/namenode
+    regexp: '^pidfile="/var/run/hadoop/hadoop-hdfs-namenode\.pid"'
```
`regexp:` is not necessary.

mitsururike commented on tests/integration/master_slave/spec/spec_helper.rb:
```diff
@@ -51,112 +53,51 @@
 
-def retry_and_sleep(options = {}, &block)
+# rubocop:disable Metrics/AbcSize,Metrics/MethodLength
+def retry_and_sleep(options = {})
```
"rspec-retry" gem should be used if it works.

mitsururike commented on tests/integration/master_slave/spec/default_spec.rb:
```diff
@@ -0,0 +1,185 @@
+    it "should be ready" do
+      response = nil
+      retry_and_sleep do
+        response = fetch "http://#{server(:namenode1).server.address}:5070/jmx"
```
`5070` is used repeatedly and should be a variable.

mitsururike commented on tests/serverspec/default_spec.rb:
```diff
@@ -0,0 +1,185 @@
+  it { should be_mode 644 }
+  it { should be_owned_by user }
+  it { should be_grouped_into group }
+  # empty
```
i do not understand what the comment means.

mitsururike commented on tests/integration/master_slave/spec/default_spec.rb:
```diff
@@ -0,0 +1,185 @@
+    end
+
+    it "should become master" do
+      result = current_server.ssh_exec("sudo -u hdfs hdfs haadmin -failover namenode1 namenode2")
```
user name `hdfs` should be a variable as it is repeatedly used and can be changed in the future.

mitsururike commented on tests/serverspec/default_spec.rb:
```diff
@@ -0,0 +1,185 @@
+
+describe file(log_dir) do
+  it { should exist }
+  it { should be_mode 775 }
```
`it { should be_directory }` should be here.
simple tests for files/directories should be done with a loop iterating an array of files/directories.

mitsururike commented on README.md:
```diff
@@ -1,126 +1,164 @@
-
-Author Information
-------------------
+| `hadoop_namenode_user` | user name of `namenode` | `{{ __hadoop_namenode_user }}` |
```
it is rather difficult to find a variable among many variables as the order looks random.
is it a historical order or is there any rule?

mitsururike commented on tests/integration/master_slave/spec/default_spec.rb:
```diff
@@ -0,0 +1,185 @@
+    end
+
+    it "should return NameNodeStatus == active" do
+      # {
```
i do not understand what the comment means.
is it for debugging or planning to add more tests of the contents?

Commit:
create the files in a loop

thanks to:
https://github.com/reallyenglish/ansible-role-hadoop-namenode/pull/14#discussion_r132096388

Commit:
use rspec-retry instead of retry_and_sleep

Commit:
Revert "create the files in a loop"

This reverts commit e3607cb7f80a47cad6ac4d57c37cafe0c4ccacff.

Commit:
Revert "Revert "create the files in a loop""

This reverts commit 7d683da05775d92b6c58200df122d8fb98bfb757.

Commit:
QA

Commit:
add rspec-retry

Commit:
[bugfix] fix third party repo, bump rubocop and QA

Commit:
QA
